### PR TITLE
[WIP] Add plagiarism check

### PR DESCRIPTION
**Pull request by juxd**
Added a plagiarism check script, to be used with [this script](https://github.com/rafaelperes/TA_CS1101S/tree/new-cadet-db)

This is a work in progress. WIll need advice on how to organise the files.

### CODE REVIEW

File: lib/cadet/jobs/autograder/grading_job.ex [hunk 1]
```diff
@@ -10,7 +10,7 @@ defmodule Cadet.Autograder.GradingJob do
   require Logger
 
   alias Cadet.Assessments.{Answer, Assessment, Question, Submission}
-  alias Cadet.Autograder.Utilities
+  alias Cadet.Autograder.{Utilities, PlagiarismChecker}
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/grading_job.ex]: Sort alphabetically

File: lib/cadet/jobs/autograder/grading_job.ex [hunk 2]
```diff
@@ -88,6 +92,14 @@ defmodule Cadet.Autograder.GradingJob do
     end
   end
 
+  def assessment_requires_plagiarism_check(assessment = %Assessment{}) do
+    Question
+    |> where(assessment_id: ^assessment.id)
+    |> select([:type])
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/grading_job.ex]: I think it's better to have this as `|> select([q], q.type)` and later on `|> Enum.any?(& &1 == :programming)` to save on memory not having to load the whole struct.

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  @plagiarism_script_path "../grader/TA_CS1101S/mosspy_submission.py"
+
+  def perform(assessment_id) when is_ecto_id(assessment_id) do
+    Logger.info("Running plagiarism check on Assessment #{assessment_id}")
+
+    script_result =
+      System.cmd("python", [@plagiarism_script_path, "--assessment_id", to_string(assessment_id)])
+
+    script_result
+    |> elem(1)
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Use pattern matching instead on line 21.

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  defp zip_results(assessment_id) do
+    System.cmd("zip", [
+      "-r",
+      "assessment_#{assessment_id}.zip",
+      "submissions/assessment#{assessment_id}/report/",
+      "submissions/assessment#{assessment_id}/assessment_report_#{assessment_id}.html"
+    ])
+
+    assessment_id
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Might wanna check whether the `zip` command fails (check the exit value)

File: lib/cadet/jobs/autograder/grading_job.ex [hunk 5]
```diff
@@ -88,6 +92,14 @@ defmodule Cadet.Autograder.GradingJob do
     end
   end
 
+  def assessment_requires_plagiarism_check(assessment = %Assessment{}) do
+    Question
+    |> where(assessment_id: ^assessment.id)
+    |> select([q], q.type)
+    |> Repo.all()
+    |> Enum.any?(&(&1 == :programming))
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/grading_job.ex]: Actually, better to not even load this into memory. Instead do:
```elixir
Question
|> where(assessment_id: ^assessment.id)
|> where(type: ^:programming)
|> Repo.aggregate(:count, :id)
|> case do
  0 -> false
  _ -> true
end
```

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+               |> Keyword.get(:bucket_name)
+
+  @plagiarism_script_path :cadet
+                          |> Application.fetch_env!(:plagiarism_check_vars)
+                          |> Keyword.get(:plagiarism_script_path)
+
+  def perform(assessment_id) when is_ecto_id(assessment_id) do
+    Logger.info("Running plagiarism check on Assessment #{assessment_id}")
+
+    {_, exit_result} =
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Refer to https://hexdocs.pm/elixir/System.html#cmd/3. Look at the typespec: actually the return tuple is `{output, exit_code}`

So perhaps what you want is `{result, 0} = System.cmd...`

It's your choice whether to pattern match exit_code with `0` or not. If you do, it's in line with Erlang's let it crash philosophy -- will be captured by sentry. Else you should instead do sth like this:
```elixir
case System.cmd(...) do
  {result, 0} ->
    ...
  _ ->
    # handle error
end
```

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      |> select([a], a.title)
+      |> Repo.one()
+
+    response =
+      @bucket_name
+      |> ExAws.S3.put_object("/reports/assessment-#{assessment_title}", File.read!(file_name))
+      |> ExAws.request!()
+
+    # if response status < 400, the transaction was successful.
+    if Map.get(response, :status_code) < 400 do
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: What does 400 represent? If it is HTTP, isn't success `200`?

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    else
+      raise inspect(response)
+    end
+  end
+
+  defp zip_results(assessment_id) do
+    {_, zip_result} =
+      System.cmd("zip", [
+        "-r",
+        "assessment_#{assessment_id}.zip",
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: You should probably put this instead the `submissions` directory so as to not dirty the root directory

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        "submissions/assessment#{assessment_id}/report/",
+        "submissions/assessment#{assessment_id}/assessment_report_#{assessment_id}.html"
+      ])
+
+    if zip_result == 0 do
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Pattern match using case instead on the `System.cmd/2`

File: lib/cadet/jobs/autograder/plagiarism_checker.ex (hunk 6 above)
Comment (juxd) [reply lib/cadet/jobs/autograder/plagiarism_checker.ex]: yeah, this looks sensible. Will it make sense to output something like a json string as the `result`, and then parse into a `Map`? might want to get multiple value results from the python script.

Comment (indocomsoft) [reply lib/cadet/jobs/autograder/plagiarism_checker.ex]: So the python script outputs json?

Comment (juxd) [reply lib/cadet/jobs/autograder/plagiarism_checker.ex]: not yet, but I can make it do so. Will probably be doing that as there are multiple pieces of info to return.

File: lib/cadet/jobs/autograder/plagiarism_checker.ex (hunk 9 above)
Comment (juxd) [reply lib/cadet/jobs/autograder/plagiarism_checker.ex]: Do i achieve this with a pipe? I can't pass `mix credo` if i do this haha.
```
 Pipe chain should start with a raw value.
```

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+         ]) do
+      {result, 0} ->
+        result
+        |> Jason.decode!()
+        |> Map.get("assessment_id")
+        |> zip_results()
+        |> store()
+
+      _ ->
+        raise "Error running script, please check again"
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: You might wanna do something like
```elixir
{output, exit_code} ->
  raise "Error running script with exit code #{exit_code}: #{output}"
```

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    end
+  end
+
+  defp store(assessment_id) when is_ecto_id(assessment_id) do
+    file_name = "submissions/assessment_#{assessment_id}.zip"
+
+    assessment_title =
+      Cadet.Assessments.Assessment
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Please alias this

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      |> where(id: ^assessment_id)
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: We prefer `Assessment |> Repo.get(id)` over `Assessment |> where(id: ^id) |> Repo.one()`

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      |> select([a], a.title)
+      |> Repo.one()
+
+    response =
+      @bucket_name
+      |> ExAws.S3.put_object("/reports/assessment-#{assessment_title}.zip", File.read!(file_name))
+      |> ExAws.request!()
+
+    if Map.get(response, :status_code) == 200 do
+      File.rm(file_name)
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Isn't this redundant?

File: lib/cadet/jobs/autograder/plagiarism_checker.ex [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+  defp zip_results(assessment_id) do
+    case System.cmd("zip", [
+           "-r",
+           "submissions/assessment_#{assessment_id}.zip",
+           "submissions/assessment#{assessment_id}/report/",
+           "submissions/assessment#{assessment_id}/assessment_report_#{assessment_id}.html"
+         ]) do
+      {_, 0} -> assessment_id
+      {_, _} -> raise "Files cannot be zipped. Please check directories."
```
Comment (indocomsoft) [inline lib/cadet/jobs/autograder/plagiarism_checker.ex]: Same as above, might wanna include output and exit code in the `raise`

**Comment by indocomsoft** [comment]
Also, travis seems to fail because environment is not set. Please resolve this as well

**Comment by indocomsoft** [comment]
Actually @juxd do you have access to push branch to this repo? We use the branch workflow so it'd be better if you can push this as a branch instead of in your own fork.

**Comment by juxd** [comment]
Will transfer this to a branch within this repo.

**Comment by indocomsoft** [comment]
Better make a new PR so I can easily check out without adding another remote

**Comment by juxd** [comment]
moved to #276

**Pull request closed without merging**